### ISSUE TITLE
TypeError: Ir is not a function or its return value is not iterable

### ISSUE THREAD

**Issue by piotr-cz** [preactjs/preact#2824]
### Reproduction

preact@10.5.5
browser: Chromium 72.0.3626 

Code below _may_ throw **TypeError: Ir is not a function or its return value is not iterable** when calling the `setState` function:

```js
import { useState, useRef } from 'preact/compat'
import { html } from 'htm/preact'

export default function Clock({
  const [date, setDate] = useState(() => new Date)
  //                      ^ This may throw an exception

  useEffect(() => {
    const timeoutIdRef = useRef()

    updateDateInInterval(timeoutIdRef)

    return () => {
      timeoutIdRef.current && window.clearTimeout(timeoutIdRef.current)
      timeoutIdRef.current = undefined
    }

    function updateDateInInterval(timeoutIdRef) {
      setDate(new Date)

      timeoutIdRef.current = window.setTimeout(
        updateDateInInterval,
        SECONDS - Date.now() % 1000,
        timeoutIdRef
      )
    }
  }, [])

  return html`
    <div>
      ${date && date.toLocaleTimeString()}
    </div>
  `
})
```

Called from `../../../node_modules/preact/dist/preact.module.js in render at line 1:8161`
Maps to [`../../../node_modules/preact/src/diff/index.js:525:13 (constructor)`](https://github.com/preactjs/preact/blob/10.5.5/src/diff/index.js#L524-L526)

```js
        return this.constructor(props, context);
```

Called from `../../../node_modules/preact/dist/preact.module.js in g at line 1:1422`
Maps to [`../../../node_modules/preact/src/component.js:218:8 (some)`](https://github.com/preactjs/preact/blob/10.5.5/src/component.js#L218-L220)

```js
                queue.some(c => {
```

### Steps to reproduce

I have no idea how to reproduce.
Error showed up only once during 30 days, possibly after implementing lazy initial state:

```diff
-  const [date, setDate] = useState(new Date)
+  const [date, setDate] = useState(() => new Date)
```

### Expected Behavior

Code produces no errors

### Actual Behavior

I've received Sentry error report with above detals

**Comment by marvinhagemeister** [comment]
One cannot nest built-in hooks. If you add `preact/debug` as the top import in you're entry file you'll get an error message for that.

```jsx
// WRONG:
useEffect(() => {
  let myVar = useRef(-1); // useRef cannot be used inside useEffect
  // ...
}, []);

// CORRECT:
useEffect(() => {
  let myVar = -1
  // ...
}, []);
```

**Comment by piotr-cz** [comment]
@marvinhagemeister Thank you!

Indeed when I've imported `preact/debug`, full error message was [Hook can only be invoked from render methods](
https://preactjs.com/guide/v10/debugging/#hook-can-only-be-invoked-from-render-methods) and I've adapted code as per your suggestion.

I'm not completely sure these two issues are related as original exception showed up just once (not on every render) but I'll assume it's resolved for now.
